Feature Visual Image Appearance

Mikescops opened the pull request:
Hello,

This is my proposal for the image appearance feature mentioned in #72 

I'm not an expert in golang at all so feel free to suggest improvements.

For now it supports the JPEG format, but I guess it's not hard to also support PNG later for instance. I just wanted to make the simplest version possible for now.

Mikescops commented:
I tested the signature validity and rendering with multiple tools.

<img width="1526" alt="image" src="https://github.com/user-attachments/assets/eff80b2a-b317-43b9-adc9-8a65634ac84e" />

Mikescops commented:
I just added an extra test to make sure incremental updates are also valid with consecutive signatures.

<img width="1515" alt="image" src="https://github.com/user-attachments/assets/fecfa972-617c-4ccf-b613-2620865ae9fa" />

Sounds to work well too.

vanbroup reviewed (commented):
Thanks! This is a great improvement.

Instead of using a Copy function, I think it would be cleaner to set a TEMP environment variable directly in the Docker development container. This would also make the resulting test files more easily accessible for debugging.

Also, a common use case for signature images is using them as a background or watermark behind text. Do you think that’s possible with the current approach? If so, could we add a test case for it?

Mikescops commented on sign/helpers.go:
- copyfile: sure I can try this approach

- background image: i need to change the way image are inserted in the PDF to support this

Mikescops commented:
@vanbroup I applied all the changes you suggested and improved the code by splitting it more into small functions

Also added a watermark option:

<img width="1521" alt="image" src="https://github.com/user-attachments/assets/324dc4e9-781e-49a2-91bd-ea73e97f6849" />

vanbroup commented on sign/appearance.go:
```diff
@@ -3,60 +3,180 @@ package sign
+	buffer.WriteString("     /F1 <<\n")
+	buffer.WriteString("       /Type /Font\n")
+	buffer.WriteString("       /Subtype /Type1\n")
+	buffer.WriteString("       /BaseFont /Times-Roman\n")
```
We’ll likely want to make this configurable in the future. Default fonts might already be supported, but using custom fonts would require embedding them into the document if they haven’t been imported yet.

That said, I think it’s fine to leave it as is for now—especially since the planned API update should make this easier to handle.

Mikescops replied:
Sure, i guess it's out of this PR scope

vanbroup reviewed (commented):
Thanks for the quick updates! 

I won’t be able to test the changes until next week, but we should be in a good position to merge them after that.

If you’re able to add support for additional image formats like PNG and GIF in the meantime, that would be great.

I’m not sure how much work it would be to also support a vector format (such as an embedded PDF), but that would be awesome and enable the creation of high quality PDF signatures!

Mikescops commented:
@vanbroup PNG support was a hard one, it requires some image manipulation to make the alpha layer work
I also added compression because otherwise the PDF size would be huge with PNGs

<img width="1103" alt="image" src="https://github.com/user-attachments/assets/5fcc9e60-182b-4849-a746-bbea8e17ae1d" />

Mikescops commented:
Concerning GIF support, I hope you never want to do that, it's ridiculously annoying to implement.

The thing is that you have to embed a color index table in the PDF (+ one for the alpha transparency) and link it back to the image object.

As for embedded PDF, I think it's doable, it doesn't seem to require anything too complex (as far as understand the spec), but I guess it will create too much confusion to happen in this PR.

vanbroup commented:
PDF would definitely provide more value than GIF.

vanbroup commented:
Have you tested JPG support after adding PNG support? I think that might have broken something.

While all files have a valid signature on both Acrobat, Foxit PDF Reader, and Edge, the following tests are **not displaying** the signature image:

- `TestSignPDFWithImage`
- `TestSignPDFWithTwoImages`
- `TestSignPDFWithWatermarkImageJPG` (text is shown)

![image](https://github.com/user-attachments/assets/33ee396a-e167-48b3-96d9-5bd65eaf78c3)

The test `TestSignPDFWithWatermarkImagePNG` seems to work fine.

vanbroup commented:
Maybe you can also add something like this to the README?

```md
## Signature Appearance with Images

You can add an image (JPG or PNG) to the visible signature appearance. This is useful for including a handwritten signature or a company logo in the signature field.

**Supported image formats:** JPG and PNG.

### Example: Signing a PDF with a visible signature and image

```go
signatureImage, err := os.ReadFile("signature.jpg")
if err != nil {
    log.Fatal(err)
}

err := sign.Sign(inputFile, outputFile, rdr, size, sign.SignData{
    Signature: sign.SignDataSignature{
        Info: sign.SignDataSignatureInfo{
            Name:        "John Doe",
            Location:    "Somewhere",
            Reason:      "Signed with image",
            ContactInfo: "None",
            Date:        time.Now().Local(),
        },
        CertType:   sign.ApprovalSignature,
        DocMDPPerm: sign.AllowFillingExistingFormFieldsAndSignaturesPerms,
    },
    Appearance: sign.Appearance{
        Visible:     true,
        LowerLeftX:  400,
        LowerLeftY:  50,
        UpperRightX: 600,
        UpperRightY: 125,
        Image:       signatureImage, // JPG or PNG image bytes
        // ImageAsWatermark: true,   // Optional: set to true to use image as watermark
    },
    DigestAlgorithm: crypto.SHA512,
    Signer:          privateKey,
    Certificate:     certificate,
})
if err != nil {
    log.Println(err)
} else {
    log.Println("Signed PDF written to " + output)
}
```
```

Mikescops commented:
Ah jeez, I'll take a look soon, probably broke something while refactoring. Too bad we can't catch that with the tests.

Mikescops commented:
@vanbroup When adding the compression to images I forgot the decode keyword for JPG, wasn't much to fix.

I updated the README.md as requested.

Mikescops commented:
Anything else before it's mergeable?

vanbroup commented:
I think we are almost there, I did apply some fixed in the main branch so it would be good to rebase your branch. 

We might need to add **FirstChar**, **LastChar**, **Widths** and **FontDescriptor** to the font dict to ensure PDF 1.7+ compatibility:

> The PostScript language names of 14 Type 1 fonts, known as the standard 14 fonts, are as follows: 
> Times-Roman, Helvetica, Courier, Symbol, Times-Bold, Helvetica-Bold, Courier-Bold, ZapfDingbats, 
> Times-Italic, Helvetica-Oblique, Courier-Oblique, Times-BoldItalic, Helvetica-BoldOblique, Courier
> BoldOblique. 
> 
> In PDF 1.0 to PDF 1.7, the **FirstChar**, **LastChar**, **Widths** and **FontDescriptor** (see “Table 109 — 
> Entries in a Type 1 font dictionary") were optional in Type 1 font dictionaries for the standard 14 fonts. 
> PDF processors supporting PDF 1.0 to PDF 1.7 files are required to have these fonts, or their font metrics and 
> suitable substitution fonts, available.

Mikescops commented:
I rebased, updated go in the devcontainer.

> We might need to add **FirstChar**, **LastChar**, **Widths** and **FontDescriptor** to the font dict to ensure PDF 1.7+ compatibility

I gave it a try, I'm not sure how to confirm it works well.
Took metrics for the Times-Roman font here: http://download.macromedia.com/pub/developer/opentype/tech-notes/Core14_AFMs.zip
And doc here: https://adobe-type-tools.github.io/font-tech-notes/pdfs/5004.AFM_Spec.pdf

Extract from T-R metrics file:
```
FontName Times-Roman
FullName Times Roman
FamilyName Times
Weight Roman
ItalicAngle 0
IsFixedPitch false
CharacterSet ExtendedRoman
FontBBox -168 -218 1000 898 
UnderlinePosition -100
UnderlineThickness 50
Version 002.000
Notice Copyright (c) 1985, 1987, 1989, 1990, 1993, 1997 Adobe Systems Incorporated.  All Rights Reserved.Times is a trademark of Linotype-Hell AG and/or its subsidiaries.
EncodingScheme AdobeStandardEncoding
CapHeight 662
XHeight 450
Ascender 683
Descender -217
StdHW 28
StdVW 84
StartCharMetrics 315
```

shengzhi commented:
hello,
Thank you both for your great work!
Could you let me know when the feature will be available?
